File: macclean/cleaners/update.py
import shutil
from pathlib import Path
import click
from rich.console import Console
from rich.panel import Panel
from macclean.core.utils import run_as_user

console = Console()


@click.command()
@click.option("--brew/--no-brew", default=True, show_default=True)
@click.option("--pip/--no-pip", "do_pip", default=True, show_default=True)
@click.option("--npm/--no-npm", default=True, show_default=True)
@click.pass_context
def cmd(ctx, brew: bool, do_pip: bool, npm: bool):
    """Upgrade all outdated packages across Homebrew, pip, and npm."""
    console.print(Panel("[bold cyan]Package Updates[/]"))

    if brew and shutil.which("brew"):
        console.print("\n[bold]Homebrew[/]")
        out, code = run_as_user(["brew", "upgrade"], timeout=300)
        if code == 0:
            console.print("  [green]+[/] brew upgrade complete")
        else:
            console.print(f"  [yellow]![/] brew upgrade: {out[:300]}")

    if do_pip:
        console.print("\n[bold]pip[/]")
        out, _ = run_as_user(["pip", "list", "--outdated", "--format=freeze"])
        pkgs = [line.split("==")[0] for line in out.splitlines() if "==" in line]
        if pkgs:
            _, code = run_as_user(["pip", "install", "--upgrade"] + pkgs, timeout=120)
            if code == 0:
                console.print(f"  [green]+[/] Upgraded {len(pkgs)} pip package(s)")
            else:
                console.print(f"  [yellow]![/] pip upgrade failed")
        else:
            console.print("  [green]+[/] All pip packages up to date")

    if npm and shutil.which("npm"):
        console.print("\n[bold]npm[/]")
        out, code = run_as_user(["npm", "update", "-g"], timeout=120)
        if code == 0:
            console.print("  [green]+[/] npm global packages updated")
        else:
            console.print(f"  [yellow]![/] npm update: {out[:200]}")

File: src/cleaners/update.rs
use anyhow::Result;
use crate::core::cmd::run_as_user;
use crate::ui::{print_ok, print_warn};

pub fn run(do_brew: bool, do_pip: bool, do_npm: bool) -> Result<()> {
    // ── Homebrew ──────────────────────────────────────────────────────────────
    if do_brew {
        println!("  Updating Homebrew packages...");
        let r = run_as_user(&["brew", "upgrade"]);
        if r.success() {
            print_ok("brew upgrade completed.");
        } else {
            print_warn(&format!(
                "brew upgrade: {}",
                &r.output[..r.output.len().min(300)]
            ));
        }
    }

    // ── pip ───────────────────────────────────────────────────────────────────
    if do_pip {
        println!("  Checking outdated pip packages...");
        let list_r = run_as_user(&["pip", "list", "--outdated", "--format=freeze"]);
        if !list_r.success() {
            print_warn("pip not available or failed to list outdated packages.");
        } else {
            let packages: Vec<&str> = list_r
                .output
                .lines()
                .filter_map(|l| l.split("==").next())
                .filter(|s| !s.trim().is_empty())
                .collect();

            if packages.is_empty() {
                print_ok("All pip packages are up to date.");
            } else {
                println!("  Upgrading {} pip package(s)...", packages.len());
                let mut args = vec!["pip", "install", "--upgrade"];
                args.extend_from_slice(&packages);
                let upgrade_r = run_as_user(&args);
                if upgrade_r.success() {
                    print_ok("pip upgrade completed.");
                } else {
                    print_warn(&format!(
                        "pip upgrade: {}",
                        &upgrade_r.output[..upgrade_r.output.len().min(300)]
                    ));
                }
            }
        }
    }

    // ── npm ───────────────────────────────────────────────────────────────────
    if do_npm {
        println!("  Updating global npm packages...");
        let r = run_as_user(&["npm", "update", "-g"]);
        if r.success() {
            print_ok("npm update -g completed.");
        } else {
            print_warn(&format!(
                "npm update: {}",
                &r.output[..r.output.len().min(300)]
            ));
        }
    }

    Ok(())
}
